### PR TITLE
Train performances improvement

### PR DESCRIPTION
**Pull request by ocourtin**
While training GPU spend lot of time idle, mainly for two reasons:
- Lot of time spend with CPU computation (scikit metrics)
- No enough workers to feed data 

This PR fix both.

And as a result, using several GPU for training become meaningful.

### CODE REVIEW

**Comment by ocourtin** [comment]
Indeed no longer need scikit stuff.

About predict, it's a bit better, but guess could be improve again (GPU-Utilisation and multi GPU).
Will give a look, but not on my top list for now.

Launch black, without option, on metrics.py file 
(if i do so on train.py it would format code i did'nt touch...)
Dunno if you use black with specific options

File: robosat/metrics.py [hunk 1]
```diff
@@ -27,7 +27,14 @@ def add(self, actual, predicted):
...
 
-        matrix = sklearn.metrics.confusion_matrix(actual, predicted, labels=self.labels)
+        confusion = predicted / actual
+
+        tn = torch.sum(torch.isnan(confusion)).item()
+        fn = torch.sum(confusion == float("inf")).item()
+        fp = torch.sum(confusion == 0).item()
+        tp = torch.sum(confusion == 1).item()
+
+        matrix = np.array([[tn, fn], [fp, tp]])
```
Comment (daniel-j-h) [inline robosat/metrics.py]: Did you run this and compared it against the sklearn confusion matrix side by side?

File: robosat/tools/train.py [hunk 2]
```diff
@@ -143,9 +144,9 @@ def train(loader, num_classes, device, net, optimizer, criterion):
         running_loss += loss.item()
 
         for mask, output in zip(masks, outputs):
-            mask = mask.data.cpu().numpy()
-            prediction = output.data.max(0)[1].cpu().numpy()
-            iou.add(mask.ravel(), prediction.ravel())
+            prediction = output.detach()
+            prediction.requires_grad = False
```
Comment (daniel-j-h) [inline robosat/tools/train.py]: Why detach and disabling gradients here? How is this different than moving the data to numpy arrays? Why do we need this?

File: robosat/tools/train.py [hunk 3]
```diff
@@ -179,16 +180,14 @@ def validate(loader, num_classes, device, net, criterion):
         running_loss += loss.item()
 
         for mask, output in zip(masks, outputs):
-            mask = mask.data.cpu().numpy()
-            prediction = output.data.max(0)[1].cpu().numpy()
-            iou.add(mask.ravel(), prediction.ravel())
+            iou.add(mask.float(), output.max(0)[1].float())
```
Comment (daniel-j-h) [inline robosat/tools/train.py]: Why do we not have to detach here like above? I don't think I understand how detach/nograd interacts and what the purpose of it is :thinking:

File: robosat/tools/train.py (hunk 2 above)
Comment (ocourtin) [reply robosat/tools/train.py]: It's not a loss, it's a metrics so why would you use grads there ?

And it's different from numpy, because it's on GPU and not on CPU.
So no need to move data from one device to an another.

File: robosat/tools/train.py (hunk 3 above)
Comment (ocourtin) [reply robosat/tools/train.py]: No need to detach there because you already have a @no_grad  on validate function

**Comment by ocourtin** [comment]
My environnement is not used only for robosat, so pip freeze return quite others pip modules already installed. So not very helpful.

But if i look only on deps who are explicitly imported later in code, it could be reduce as:

attrs==18.1.0
Flask==1.0.2
geojson==2.4.0
matplotlib==2.2.2
mercantile==1.0.4
numpy==1.14.4
osmium==2.14.1
py==1.5.3
pyproj==1.9.5.1
rasterio==1.0b1
requests==2.18.4
Rtree==0.8.3
Shapely==1.6.4.post1
supermercado==0.0.5
toml==0.9.4
torch==0.4.0
torchvision==0.2.1
tqdm==4.23.4




Thanks for the black options setting.
9d9baa4 commit was already OK for both metrics.py and train.py

**Comment by ocourtin** [comment]
Hum meant bc27447

File: robosat/tools/train.py (hunk 2 above)
Comment (daniel-j-h) [reply robosat/tools/train.py]: Hm the docs say the result of detach will never require gradient. Do we need to explicitly set the requires grad attribute to false here then?

https://pytorch.org/docs/stable/autograd.html#torch.Tensor.detach

File: robosat/tools/train.py (hunk 3 above)
Comment (daniel-j-h) [reply robosat/tools/train.py]: :+1:

File: robosat/tools/train.py (hunk 2 above)
Comment (ocourtin) [reply robosat/tools/train.py]: You're right. No need to set these already implicit statement.
Fixed as bfeab0e

Thanks for you code review !

**Comment by daniel-j-h** [comment]
Thanks again, this is awesome! :tada:
